modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/model/listener/HeadlessBuilderObjectEntryModelListener.java

```diff
@@ -0,0 +1,154 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.headless.builder.internal.model.listener;
+
+import com.liferay.headless.builder.application.APIApplication;
+import com.liferay.headless.builder.application.provider.APIApplicationProvider;
+import com.liferay.headless.builder.application.publisher.APIApplicationPublisher;
+import com.liferay.object.model.ObjectDefinition;
+import com.liferay.object.model.ObjectEntry;
+import com.liferay.object.service.ObjectDefinitionLocalService;
+import com.liferay.object.service.ObjectEntryLocalService;
+import com.liferay.portal.kernel.exception.ModelListenerException;
+import com.liferay.portal.kernel.model.BaseModelListener;
+import com.liferay.portal.kernel.model.ModelListener;
+import com.liferay.portal.kernel.transaction.TransactionCommitCallbackUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.io.Serializable;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Sergio Jiménez del Coso
+ */
+@Component(service = ModelListener.class)
+public class HeadlessBuilderObjectEntryModelListener
+	extends BaseModelListener<ObjectEntry> {
+
+	@Override
+	public void onAfterCreate(ObjectEntry objectEntry)
+		throws ModelListenerException {
+
+		_schedulePublication(objectEntry);
+	}
+
+	@Override
+	public void onAfterUpdate(
+			ObjectEntry originalObjectEntry, ObjectEntry objectEntry)
+		throws ModelListenerException {
+
+		_schedulePublication(objectEntry);
+	}
+
+	private long _getAPIApplicationId(ObjectEntry objectEntry) {
+		if (_isObjectDefinition("L_API_APPLICATION", objectEntry)) {
+			return objectEntry.getObjectEntryId();
+		}
+		else if (_isObjectDefinition("L_API_SCHEMA", objectEntry)) {
+			Map<String, Serializable> values = objectEntry.getValues();
+
+			return (long)values.get(
+				"r_apiApplicationToAPISchemas_c_apiApplicationId");
+		}
+		else if (_isObjectDefinition("L_API_ENDPOINT", objectEntry)) {
+			Map<String, Serializable> values = objectEntry.getValues();
+
+			return (long)values.get(
+				"r_apiApplicationToAPIEndpoints_c_apiApplicationId");
+		}
+
+		throw new UnsupportedOperationException("Invalid object definition");
+	}
+
+	private boolean _isObjectDefinition(
+		String objectDefinitionExternalReferenceCode, ObjectEntry objectEntry) {
+
+		ObjectDefinition objectDefinition =
+			_objectDefinitionLocalService.fetchObjectDefinition(
+				objectEntry.getObjectDefinitionId());
+
+		return Objects.equals(
+			objectDefinition.getExternalReferenceCode(),
+			objectDefinitionExternalReferenceCode);
+	}
+
+	private void _schedulePublication(ObjectEntry objectEntry) {
```

Comment:
Should we be aware of concurrent access to this method? I think someone could unpublish and publish an ApiApplication in different threads, maybe it could end by accessing to this method concurrently, and it could make an ApiApplication be published after unpublished or viceversa. What do you think @4lejandrito ? maybe annotating with `synchronized` would avoid that race condition

Reply:
I thought about this but not enough. I would use synchronized too, but I don't think Brian will merge it. There are very few cases of synchronized in the codebase... we can talk about this offline.

Reply:
@sergiojimcos we can add a separate commit to use the synchronized keyword.

Reply:
https://github.com/liferay-headless/liferay-portal/pull/1321/commits/deef14960558400f8134f82b7f00c926edcfb477